adaptor.go

```diff
@@ -0,0 +1,138 @@
+package jwtverifier
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/lestrrat-go/jwx/jwk"
+	"github.com/lestrrat-go/jwx/jws"
+	"github.com/patrickmn/go-cache"
+)
+
+type Adaptor interface {
+	Decode(jwt string) (map[string]interface{}, error)
+}
+
+type JWXAdaptor struct {
+	jwksUri     string
+	jwkKeyCache *cache.Cache
+}
+
+func NewJWXAdaptor(metaDataUrl string) (Adaptor, error) {
+	jwksUri, err := getJWKSURI(metaDataUrl)
+	if err != nil {
+		return nil, fmt.Errorf("failed go get jwks uri: %w", err)
+	}
+
+	return &JWXAdaptor{
+		jwksUri:     jwksUri,
+		jwkKeyCache: cache.New(24*time.Hour, time.Hour),
+	}, nil
+}
+
+func (a *JWXAdaptor) Decode(jwt string) (map[string]interface{}, error) {
+	kid, err := getKid(jwt)
+	if err != nil {
+		return nil, err
+	}
+
+	jwkKey, err := a.getJwkKey(kid)
+	if err != nil {
+		return nil, err
+	}
+
+	token, err := jws.VerifyWithJWK([]byte(jwt), jwkKey)
+	if err != nil {
+		return nil, err
+	}
+
+	var claims map[string]interface{}
+	if err = json.Unmarshal(token, &claims); err != nil {
+		return claims, err
+	}
+
+	return claims, nil
+
+}
+
+func (a *JWXAdaptor) getJwkKey(kid string) (jwk.Key, error) {
+	if x, found := a.jwkKeyCache.Get(kid); found {
+		switch v := x.(type) {
+		case error:
+			return nil, v
+		case jwk.Key:
+			return v, nil
+		default:
+			return nil, errors.New("unknown value found in jwkKeyCache")
+		}
+	}
+
+	jwkSet, err := jwk.FetchHTTP(a.jwksUri)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: This probably needs to be just the key...
+	keys := jwkSet.LookupKeyID(kid)
+	if len(keys) < 1 {
+		a.jwkKeyCache.SetDefault(kid, ErrJWKNotFound)
```

Comment:
https://developer.okta.com/docs/reference/api/oidc/#keys
>     Note: The information returned from this endpoint could lag slightly, but will eventually be up-to-date.

We need to be careful here. Imagine this scenario:
1. Key rotation happens at 15:00
2. Until 15:02 `/v1/keys` endpoint returns stale information about old keys
3. At 15:01 client gets a new authorization token and immediately makes a request to the backed. Backend does a `/v1/keys` lookup, gets only the stale keys and doesn't let the client through.
4. At 15:03 client retries request to the backend. `/v1/keys` now returns correct information, problem is that backend won't make a request to `/v1/keys` for next 24h because it cached `ErrJWKNotFound`. As a result all clients will be denied access for next 24h.

Reply:
Seems like a big flaw on the backend, no? Doesn't this mean all users who are given a token that was signed with the new key will be unauthenticated during that period of time? So every time keys are rotated, many of the users who got new tokens will get 401s. How could that be acceptable?

Reply:
> Doesn't this mean all users who are given a token that was signed with the new key will be unauthenticated during that period of time?

Yes unless you go out of your way to add a workaround for that. Possible workarounds:
1. use [slow introspect endpoint](https://developer.okta.com/docs/reference/api/oidc/#introspect) during the transition period
2. use Manual "Signing Key Rotation" setting
3. use Okta Resources API to retrieve information about the authorization server - check "credentials.signing.kid" and "credentials.signing.lastRotated"/nextRotation

Reply:
Is this an okta specific problem? Is there any plan to fix this?

Reply:
> Is this an okta specific problem?

yes

Reply:
I think we are using the standard okta login flow (web based). Does that perform the "slow introspection" automatically?